File: XORCipher/XORCipher.cs
namespace XORCipher;

public class XorCipher
{
    public static string GetRandomKey(int key, int length)
    {
        var gamma = string.Empty;
        var rnd = new Random(key);
        for (int i = 0; i < length; i++) gamma += (char)rnd.Next(35, 126);
        return gamma;
    }
    private string GetRepeatKey(string s, int n)
    {
        var r = s;
        while (r.Length < n) r += r;
        return r.Substring(0, n);
    }

    private string Cipher(string text, string secretKey)
    {
        var currentKey = GetRepeatKey(secretKey, text.Length);
        var res = string.Empty;
        for (var i = 0; i < text.Length; i++) res += (char)(text[i] ^ currentKey[i]);
        return res;
    }

    public string Encrypt(string plainText, string password) => Cipher(plainText, password);
    
    public string Decrypt(string encryptedText, string password) => Cipher(encryptedText, password);
}

File: XORCipher FS/XORCipher.fs
namespace XORCipher_FS
open System
open System.Text

type public XORCipher() =
    static member GetRandomKey(key: int, length:int) =
        let gamma = StringBuilder()
        let rnd = Random(key)
        for i = 0 to length do gamma.Append(string(char(rnd.Next(35, 126)))) |> ignore
        gamma.ToString()
    
    member private this.GetRepeatKey(s:string) (n:int) =
        let r = StringBuilder(s)
        while r.Length < n do r.Append(r)
        r.ToString().Substring(0, n)
    
    member private this.Cipher(text: string) (secretKey: string) =
        let currentKey = this.GetRepeatKey secretKey text.Length
        let res = StringBuilder() 
        for i = 0 to text.Length - 1 do res.Append(string(char(int(text[i]) ^^^ int(currentKey[i])))) |> ignore
        res.ToString()
        
    member public this.Encrypt(plainText: string, password: string) = this.Cipher plainText password
        
    member public this.Decrypt(encryptedText: string, password: string) = this.Cipher encryptedText password